plugins/modules/ipacert.py

```diff
@@ -0,0 +1,376 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+# Authors:
+#   Sam Morris <sam@robots.org.uk>
+#
+# Copyright (C) 2021 Red Hat
+# see file 'COPYING' for use and warranty information
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from __future__ import (absolute_import, division, print_function)
+
+__metaclass__ = type
+
+ANSIBLE_METADATA = {
+    "metadata_version": "1.0",
+    "supported_by": "community",
+    "status": ["preview"],
+}
+
+DOCUMENTATION = """
+---
+module: ipacert
+short description: Manage FreeIPA certificates
+description: Request, revoke, hold and release FreeIPA certificates
+extends_documentation_fragment:
+  - ipamodule_base_docs
+options:
+  csr:
+    description:
+      X509 certificate signing request, in RFC 7468 PEM encoding.
+      Use with I(state=requested).
+    type: str
+  principal:
+    description:
+      Principal for whom the certificate is being issued.
+      Use with I(state=requested).
+      For a service certificate, use C(HTTP/www.example.com).
+      For a host certificate, use C(host/foo.example.com).
+      For a user certificate, use just the username, C(alice).
+    type: str
+  add_principal:
+    description:
+      Use with I(state=requested).
+      If I(principal) doesn't exist, add it.
+    type: bool
+  ca:
+    description:
+      Request the certificate from the named lightweight sub-CA.
+      If not specified, the default FreeIPA CA will be used.
+    type: str
+  serial_number:
+    description:
+      Certificate serial number.
+      Use with I(state=held), I(state=released), I(state=revoked).
+      Can be in decimal (C(12345)) or hexadecimal (C(0xa1b2)).
+    type: int
+  profile:
+    description:
+      Issue the certificate with the named certificate profile.
+      Use with I(state=requested).
+      If not specified, the default certificate profile (suitable for hosts and services) will be used.
+    type: str
+  reason:
+    description:
+      Reason the certificate is being revoked.
+      Use with I(state=revoked).
+    default: unspecified
+    type: str
+    choices: ["unspecified", "keyCompromise", "cACompromise", "affiliationChanged", "superseded", "cessationOfOperation", "certificateHold", "privilegeWithdrawn", "aACompromise"]
+  state:
+    description: The state to ensure.
+    choices: ["requested", "held", "released", "revoked"]
```

Comment:
As the implementation allows retrieval of the certificate, it should have a state `retrieved` to retrieve certificates by `serial number`. The parameters allowed in this case are `serial_number`, `chain`, and `certificate_out`.